server/plugin.go

```diff
@@ -0,0 +1,112 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"sync"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/pkg/errors"
+	"github.com/robfig/cron"
+)
+
+// Plugin implements the interface expected by the Mattermost server to communicate between the server and plugin processes.
+type Plugin struct {
+	plugin.MattermostPlugin
+
+	// configurationLock synchronizes access to the configuration.
+	configurationLock sync.RWMutex
+
+	// configuration is the active plugin configuration. Consult getConfiguration and
+	// setConfiguration for usage.
+	configuration *configuration
+
+	// a job for pre-calculating channel recommendations for users.
+	preCalcJob    *cron.Cron
+	preCalcPeriod string
+	botUserID     string
+}
+
+// ServeHTTP demonstrates a plugin that handles HTTP requests by greeting the world.
+func (p *Plugin) ServeHTTP(c *plugin.Context, w http.ResponseWriter, r *http.Request) {
+	fmt.Fprint(w, "Hello, world!")
+}
+
+// See https://developers.mattermost.com/extend/plugins/server/reference/
+
+type readFile func(filename string) ([]byte, error)
+
+func (p *Plugin) setupBot(reader readFile) error {
+	botID, err := p.Helpers.EnsureBot(&model.Bot{
+		Username:    "suggestions",
+		DisplayName: "Suggestions",
+		Description: "Created by the Suggestions plugin.",
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to ensure github bot")
+	}
+	p.botUserID = botID
+	bundlePath, err := p.API.GetBundlePath()
+	if err != nil {
+		return errors.Wrap(err, "couldn't get bundle path")
+	}
+
+	profileImage, err := reader(filepath.Join(bundlePath, "assets", "profile.jpeg"))
+	if err != nil {
+		return errors.Wrap(err, "couldn't read profile image")
+	}
+
+	appErr := p.API.SetProfileImage(botID, profileImage)
+	if appErr != nil {
+		return errors.Wrap(appErr, "couldn't set profile image")
+	}
+	return nil
+}
+
+func (p *Plugin) startPrecalcJob() error {
```

Comment:
Unrelated to your plugin, but in an HA environment, each plugin instance (running on each server) will spawn the same job. We need to fix the tooling to make this experience better.